commit e278c5a90368408bd191743e7c6f978f068f4b8d
Author: Christian Lamparter <chunkeey@googlemail.com>
Date:   Fri Sep 17 22:22:50 2010 +0200

    carl9170: fix noise dBm conversion
    
    Ever since carl9170 gained support to read the noisefloor,
    the reported noisefloor level was pretty poor.
    
    Initially I assumed that something was wrong in the PHY
    setup and it would be impossible to fix without any
    guidances. But this was not the case. In fact the nf
    readings were correct and the thing that was broken
    was the "simple" sign extension code!
    
    Signed-off-by: Christian Lamparter <chunkeey@googlemail.com>
    Signed-off-by: John W. Linville <linville@tuxdriver.com>

--- a/drivers/net/wireless/ath/carl9170/phy.c
+++ b/drivers/net/wireless/ath/carl9170/phy.c
@@ -1558,9 +1558,9 @@ static int carl9170_set_power_cal(struct
 static int carl9170_calc_noise_dbm(u32 raw_noise)
 {
 	if (raw_noise & 0x100)
-		return ~((raw_noise & 0x0ff) >> 1);
+		return ~0x1ff | raw_noise;
 	else
-		return (raw_noise & 0xff) >> 1;
+		return raw_noise;
 }
 
 int carl9170_get_noisefloor(struct ar9170 *ar)
